### PR TITLE
[Feature] Implement a new travel plan persistence index

### PR DESCRIPTION
**Pull request by timwangmusic**
## Description
Clearly describe changes in your PR, new feature or bug fixes.

## Solution
* [Design Doc](https://docs.google.com/document/d/15JFauy-u5L1GitJLLJWjU2MhtrzvQfP2K7Gqvpj-FeA/edit?usp=sharing)

## Testing
- [x] Integration testing on Heroku staging
- [x] Added new unit tests

## Checks
- [x] Have you removed commented code?
- [x] Have you used gofmt to format your code?

### CODE REVIEW

**Review by rwangsc18 — COMMENTED** [review]
Overall look good to me, left some comments. Let me know what you think.

File: iowrappers/redis_client.go [hunk 1]
```diff
@@ -478,63 +478,84 @@ type PlanningSolutionsResponse struct {
...
+
+	parts = append(parts, interval.Start.ToString())
+	parts = append(parts, interval.End.ToString())
+
+	if int(weekday) > 6 || int(weekday) < 0 {
+		return "", fmt.Errorf("weekday falls between 0 and 6 inclusive, got %s", weekday.String())
+	}
+
+	parts = append(parts, weekday.String())
+	return strings.Join(parts, "-"), nil
```
[the remark this answers is not in the document]
Comment (timwangmusic) [reply iowrappers/redis_client.go:538]: I think using dashes mixed with underscores make it more human readable, wdyt?

File: iowrappers/redis_client.go [hunk 2]
```diff
@@ -478,63 +478,84 @@ type PlanningSolutionsResponse struct {
...
 	region = strings.ReplaceAll(strings.ToLower(region), " ", "_")
 	city = strings.ReplaceAll(strings.ToLower(city), " ", "_")
 
-	redisFieldKey := strings.ToLower(strings.Join([]string{TravelPlansRedisCacheKeyPrefix, country, region, city, radius, strconv.Itoa(int(req.Weekday)), strconv.Itoa(int(req.PriceLevel)), planIndexStr}, ":"))
+	redisFieldKey := strings.ToLower(strings.Join([]string{TravelPlansRedisCacheKeyPrefix, country, region, city, strconv.Itoa(int(req.PriceLevel)), slotsIndex}, ":"))
 	return redisFieldKey, nil
 }
 
-func (r *RedisClient) SavePlanningSolutions(context context.Context, request *PlanningSolutionsCacheRequest, response *PlanningSolutionsResponse) error {
+func (r *RedisClient) SavePlanningSolutions(context context.Context, request *PlanningSolutionsSaveRequest) error {
```
[the remark this answers is not in the document]
Comment (timwangmusic) [reply iowrappers/redis_client.go:556]: The cache request should contain the meta data and the data we want to cache, so I think the new structure makes more sense.

File: iowrappers/redis_client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rwangsc18) [reply iowrappers/redis_client.go:538]: Ah, I didn't realize that "-" is "_" are used differently in the new key. It makes sense to have both delimiters then.

File: iowrappers/redis_client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rwangsc18) [reply iowrappers/redis_client.go:556]: icic, thanks!

**Comment by rwangsc18** [comment]
LGTM

**Review by rwangsc18 — APPROVED** [review]

**Pull request merged**